c10/core/GeneratorImpl.h

```diff
@@ -81,18 +80,9 @@ struct C10_API GeneratorImpl : public c10::intrusive_ptr_target {
 
   DispatchKeySet key_set() const { return key_set_; }
 
-  inline void set_pyobj(PyObject* pyobj) noexcept {
-    pyobj_ = pyobj;
-  }
-
-  inline PyObject* pyobj() const noexcept {
-    return pyobj_;
-  }
-
   protected:
     Device device_;
     DispatchKeySet key_set_;
-    PyObject* pyobj_ = nullptr;
```

Comment:
This looks like a pessimization. If you don't cache the pyobj in the Generator struct, then pybind11 has to manually maintain object identity on its own side, and the only way to do that is with a giant hash table mapping, which is slower than just directly storing it on the object in question. This should show up in the profiles.

Reply:
I agree that pybind11 will need to do lookup in an object identity hash table, but isn't that cost always present with every class you bind with pybind11? If so having `pyobj_` doesn't make sense, because even though you can avoid the lookup, pybind11 will still insert the Python object for the `Generator` into that lookup table. You are only going to end up with some minor speedups at the cost of more memory consumption, and since `Generator` is not that frequently used (here I mean not at the scale of `torch.Tensor`, which we probably never want to bind with pybind11) I prefer not to keep `pyobj_` on `Generator`.

Reply:
You wouldn't have the problem you're fixing in https://github.com/pytorch/pytorch/pull/49771#discussion_r554207279 if you used the `pyobj_` field 👀

But yes, everything bound in pybind11 that uses pybind's automatic identity preservation will be slower this way. You might call this a microoptimization... but the microoptimization is easy and non-invasive; really, the burden of proof should be on *not* doing it. (And that's not considering the correctness problem you have)

Reply:
Yeah I agree this will be helpful for performance, but another reason I oppose having `pyobj_` is that it violates the dependency between c10/ATen, PyTorch and `libpython`. Ideally `at::Generator` and `c10::GeneratorImpl` should not depend on `libpython` (instead that should be handled by the PyTorch bindings), however having the `pyobj_` field defeats that purpose.

That said, if you are really concerned about the performance, I have a better idea which does not depends on `pyobj_`. While pybind11 maintains a mapping between Python and C++ objects, it is only looked up when we are casting a pointer or a reference to a Python object. AFAIK, this is only triggered for `Generator.set_state` and `identity` (both return the Generator itself). If we change the generator argument from `Generator` to `py::object` for them, then we should be able to avoid the lookup and remove the overhead.

Reply:
In terms of build dependency, we have the same in TensorImpl: https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1045 - the reason it works is that you can forward-declare PyObject in libtorch but don't really depend on any python headers.

We can do the same for Generator and return pre-looked-up py::object as you noted in your comment. You probably would need to register custom wrapper with pybind to "remove" the py::object identity from the object when the last reference gets deleted (I'm not sure how exactly to do it, checking pybind sources might be useful)